src/c_api/c_api_symbolic.cc

```diff
@@ -1199,3 +1200,73 @@ int MXShallowCopySymbol(SymbolHandle src, SymbolHandle* out) {
   *out = out_sym;
   API_END_HANDLE_ERROR(delete out_sym);
 }
+
+int MXOptimizeForBackend(SymbolHandle sym_handle,
+                         const char* backend_name,
+                         SymbolHandle* ret_sym_handle,
+                         const mx_uint len,
+                         NDArrayHandle* in_args_handle,
+                         const mx_uint num_options,
+                         const char** keys,
+                         const char** vals) {
+  nnvm::Symbol *s = new nnvm::Symbol();
+  API_BEGIN();
+  nnvm::Symbol *sym = static_cast<nnvm::Symbol *>(sym_handle);
+  *s = sym->Copy();
+  nnvm::Graph g = Symbol2Graph(*s);
+  if (len) {
+    NDArray **in_args_ptr = reinterpret_cast<NDArray**>(in_args_handle);
+    Context default_ctx = in_args_ptr[0]->ctx();
+    mxnet::ShapeVector arg_shapes(len);
+    nnvm::DTypeVector arg_dtypes(len);
+    StorageTypeVector arg_stypes(len);
+    for (mx_uint i = 0; i < len; i++) {
+      const auto &in_arg = *(in_args_ptr[i]);
+      arg_shapes[i] = in_arg.shape();
+      arg_dtypes[i] = in_arg.dtype();
+      arg_stypes[i] = in_arg.storage_type();
+      CHECK(in_arg.ctx() == default_ctx)
+          << "args[" << i << "] is on context: " << in_arg.ctx()
+          << ", whereas args[0] is on context: " << default_ctx
+          << ". All args must be on the same context.";
+    }
+    const auto& indexed_graph = g.indexed_graph();
+    const auto num_forward_inputs = indexed_graph.input_nodes().size();
+    g.attrs["context"] = std::make_shared<nnvm::any>(
+        exec::ContextVector(indexed_graph.num_nodes(), default_ctx));
+    // infer shapes
+    g = exec::InferShape(std::move(g), std::move(arg_shapes), "__shape__");
```

Comment:
Shall we call `InferForwardAttrs` instead? Then we can maintain single copy of this functionality.
https://github.com/apache/incubator-mxnet/blob/d8b6e47b91a737d29c0589bda8640aee69d8599b/src/executor/graph_executor.cc#L1606

Reply:
I think the differences are minor, and they won't cause problems. Overall, I think we need to unify the build subgraph process to make it behaves all the same either from bind API or optimize_for API.

Reply:
That's fine with me. Alternatively, it's better to add some comments to mention this, if other developer changed `InferForwardAttrs` for the subgraph process in `bind` API, probably that change need to apply to `optimize_for` as well.